Visualize 3D models locally without relying on sketchfab iframe

luistoptal opened the pull request:
# Pull request for issue: #2054

This is a pull request for the following functionalities:

* Add three.js widget to visualize dataset 3D models without relying on sketchfab. This module is selfcontained, can be reused anywhere on the site and takes a list of external links to 3D models as data input

## How to test?

The following steps describe how I test this widget locally (if you know of a better way, please be my guest and I would love to know 🙂):

- download the models from the following urls to your computer (or feel free to use any files you want as long as they are LAS, STL, PLY or OBJ)
  - https://sketchfab.com/3d-models/leaf-01-point-cloud-5acb196dd274419c8e3c614e55d2cf8b (LAS file)
  - https://sketchfab.com/3d-models/beet-03-c63d100fb76042ecab2ec87f73da323e (STL file)
  - https://sketchfab.com/3d-models/alimentary-system-of-xenopus-laevis-nf-66-30b2ebcc9b364568a7a04ed87e81e6fd (PLY file)
  - https://sketchfab.com/3d-models/globigerina-bulloides-geo-b8502-865cm-shell-4-b7b47f84c7f84623b7036f585d89770a (OBJ file)
- Manually add the files to the assets folder, e.g: /assets/test-models/*_YOUR_FILES_*
- http://gigadb.gigasciencejournal.com/adminExternalLink/create add the files one by one as external links to any one of the published datasets (e.g. 100006). In the url field you should add something like this: /assets/test-models/__filename__, in External Link Type select "3d Models"
- Navigate to the dataset page, e.g.: http://gigadb.gigasciencejournal.com/dataset/100006
- You should see the "3D models" tab. When you open it, you should see the models you added in the dropdown
- The initial state of the viewer should be no loaded modal, and a visible play button
- when you click play, a model should be loaded and you should be able to interact with it
- the "help" and "fullscreen" buttons should work as intended, either with mouse or keyboard
- if you toggle fullscreen and / or resize the window, the viewer should be fully responsive and the model should remain visible
- if you select another model in the selector, the model should be immediately loaded

![image](https://github.com/user-attachments/assets/9c5e8131-df8e-4ce7-ac06-4bf94774bb11)


You can test error states if you repeat the above process by adding a file that is not a valid 3d model or not a LAS, STL, PLY or OBJ file (or you can just add a `throw new Error('Ooops')` in the line before `await loadModel(file);`)

Example of error message:

![image](https://github.com/user-attachments/assets/9b7b0687-4c04-4cbc-b890-2dca009cdba7)

- it would be critical to test in a production-like environment with files / external links mimicking the production environment

## How have functionalities been implemented?

3D models were previously visualized within an iframe taking a sketchfab url as input. This PR gets rid of the sketchfab iframe and replaces it with a JS widget that makes use of the three.js library to load and display an interactive 3D model

The main requirement is to load STL and OBJ file formats, and also (less common) PLY and LAS

Different options to locally load the model were considered (refer to the original ticket for some). I decided to use three.js because:

- It is widely popular and well known, essentially being the standard way to render 3D files in JS world
- It supports many formats via built-in loaders (LAS being an exception)
- It supports / will support WebXR via plugins. Note that WebXR is an experimental API and was not considered in this PR
- For a simple use case such as rendering a model with no animations and with basic interactions it is relatively simple to implement

Other options I checked had very low or null adoption and / or did not fully cover the use case. I opted for the widely adopted solution.

The changes can be grouped as follows:

- Mainly, everything related to the model viewer is within `protected/js/model-viewer`
- The partial `protected/views/shared/_model_viewer.php` renders the html for the widget, and loads the necessary scripts;
  - three.js imports are added to the `head` as a import map. That way, the three.js library can be imported as a module in the widget scripts
  - `Yii::app()->assetManager->forceCopy = YII_DEBUG;`  makes sure the assets are rebuilt and not cached in development  (NOTE: I am unclear on whether this targets specifically local dev, is this the correct solution? If this is a blocker, this line can be removed, but I found it necessary to add this line during local dev)
  - widget scripts are added to the body
  - modelviewer is initialized
- The partial is rendered in the dataset page `protected/views/dataset/view.php`. This page contains a switch statement that renders elements such an iframe depending on the external link types. In the switch statement, the links are handled one by one. In the case of 3D models, I deleted that case from the switch statement, and passed them to the partial as `data` property. Note that I had to do this because the partial, rather than handling individual links one by one, takes them all to build a select input from which to select the model to load
- `less/modules/model-viewer.less` takes care of all the styles related to the viewer

That covers all the changes outside of the `protected/js/model-viewer` folder, now related to the widget itself

- first, the idea is to abstract implementation details away from the consuming view. I suggest looking at the jsdocs decorating the `modelViewer` function and dig deeper only if curious about implementation details
- the widget is split into two core modules
  - ui: takes care of updating the state of the UI depending on the state of the viewer (i.e.: loading states, model loaded, no model loaded, error state)
    - the UI has 4 states:
      - idle: no model is loaded, a play button to load the currently selected model (by default, the first in the select input) should be visible
      - loading: a loading spinner should be displayed while loading a model
      - error: a danger alert should be displayed at the bottom with a relevant message. Use can select another model or click "play" again
      - success: model is loaded and should be interactive
  - viewer: it uses the three.js library to render a 3D model, and takes care of all related things (setting up scene, lights, cameras, controls, loading models...)
    - It is expected that all models are in one of LAS, STL, PLY or OBJ file formats. If another format is needed, the corresponding loader for such file format should be added (protected/js/model-viewer/viewer/components/models/loader.js)
    - among other things, this module takes care of making the viewer fully responsive
    - LAS files (data point files) needed a custom implementation, with a visualization looking like a relatively crude set of points. I see sketchfab doesn't do much better, and I understand LAS files are a minority (?) so I decided not to dig more on that field
    - the "look" of the 3D models is fully customizable. I opted for a metallic effect somewhat similar to sketchfab. The visuals can be fine-tuned / changed if needed as three.js is fully customizable
- the entry point `modelViewer` connects the two modules and initializes them

## Any issues with implementation?

- I am unable to test on production or in a production-like environment
- it assumes the incoming files are defined as ExternalLinks and follow the interface protected/interfaces/DatasetExternalLinksInterface.php, if the requirement is different, for example, a different model such as protected/models/File.php, it should be relatively easy to adapt
- it assumes the site can fetch and load the incoming external links without cors issues, in short, cors is not handled in any way
- In Yii, what is the right way to access the current environment from within a .js file? For example to determine the  `isProduction` value in `protected/js/model-viewer/helpers/logger.js`

## Any changes to automated tests?

All tests pass

luistoptal committed:
feat: able to render 3d models locally

luistoptal committed:
feat: use jquery

luistoptal committed:
feat: scaffold model viewer refactor

luistoptal committed:
feat: able to visualize rendered models

luistoptal committed:
feat: model looks similar to sketchfab

luistoptal committed:
feat: adds multitud of small refactorings

luistoptal committed:
feat: load ply, obj and las models

luistoptal committed:
feat: responsive canvas and refactored ui code

luistoptal committed:
feat: improves help controls

luistoptal committed:
feat: adds fullscreen btn

luistoptal committed:
feat: refactors root ui script, colocates selectors, adds js docs

luistoptal committed:
feat: integrates model viewer into dataset page

luistoptal committed:
feat: improves a11y for model loading state

luistoptal committed:
chore: remove debug logs

luistoptal committed:
chore: up log

luistoptal committed:
chore: undoes ignored temp models folder

luistoptal committed:
feat: adds jsdocs

luistoptal committed:
chore: Merge branch 'develop' into new-feature/2054-3d-image-viewer

kencho51 commented:
Hi @luistoptal,

**prasie:** Was able to associate this https://sketchfab.com/3d-models/beet-03-c63d100fb76042ecab2ec87f73da323e stl file to dataset 100006, and display it on Chrome browser.

<img width="1755" alt="image" src="https://github.com/user-attachments/assets/cba82e2d-ecb5-4854-87e3-a3b68a0328e9">


**question:** The 3D image cannot be shown in Safari nor Firefox browsers, can you help further look into it?

### Safari
![image](https://github.com/user-attachments/assets/124ecbf3-8223-4a1f-bf8d-4a954cfb8e3b)

### Firefox
![image](https://github.com/user-attachments/assets/1958c707-4924-4a5f-a0f6-ff3802aa633c)

luistoptal commented:
@kencho51 looking at the images, it looks as if the CSS is not loaded. Because not only the 3D model is not shown but the buttons look as if the custom styles applied to them were not working. Can you try:

- re-run the ./up.sh script (but I assume you already did that)
- rebuild the .less styles using `docker-compose run --rm less` (if you run ./up.sh, it should already be done though)
- Disable Firefox and Safari browser cache and hard refresh the page

kencho51 commented:
Hi @luistoptal,


> * Disable Firefox and Safari browser cache and hard refresh the page

After hard refresh/empty the cache in safari and firefox, the 3D image can be displayed now.

kencho51 commented:
Hi @luistoptal,

**question:** Would it be better to have acceptance tests/accessibility tests as there is an extra `3D Models` tab and a 3D graph is displayed?
**suggestion:** For the 3D images source, we could put all the images in a s3 bucket and make them public, and then input the bucket url when trying to create external link, as below:
![image](https://github.com/user-attachments/assets/4f4bf32f-06b1-4bb4-bf56-1abb3cc409a2)
I have created a test bucket with 3 3D images:
1. https://s3.ap-northeast-1.wasabisys.com/test-gigadb-datasets/3d-models/100006/GeoB8502_865cm_Shell-4.obj
2. https://s3.ap-northeast-1.wasabisys.com/test-gigadb-datasets/3d-models/100006/beet_03.stl
3. https://s3.ap-northeast-1.wasabisys.com/test-gigadb-datasets/3d-models/100006/leaf_01.las

@rija, @pli888 , what do you think?

**praise:** I was able to deploy this PR to my staging, and display the 3D images from the above files stored in the s3 bucket.

luistoptal commented:
@kencho51 that would work since the viewer simply loads and displays files accessible via public links, thanks for adding these files, now they can be used to update the tests

there is one test for the 3D models tab already present, it hadn't been updated

luistoptal committed:
chore: update 3d models tab test

luistoptal committed:
feat: tests for model selector

luistoptal commented:
I extended the test that checks for the 3D Model tab

kencho51 commented:
Hi @luistoptal,

**issue:** I have further tested this PR in my production, and the 3D image actually cannot be displayed properly as below:

![image](https://github.com/user-attachments/assets/12bec221-7d6a-4b0c-a9a6-4e8c68781c0c)


and figured out that the following external scripts were blocked by the CSP:
```
Loading failed for the module with source “https://cdn.jsdelivr.net/npm/three@0.170.0/examples/jsm/loaders/STLLoader.js”. [100006](https://ec2-staging.gigadb.link/dataset/100006)
Loading failed for the module with source “https://cdn.jsdelivr.net/npm/three@0.170.0/examples/jsm/loaders/OBJLoader.js”. [100006](https://ec2-staging.gigadb.link/dataset/100006)
Loading failed for the module with source “https://cdn.jsdelivr.net/npm/three@0.170.0/examples/jsm/loaders/PLYLoader.js”. [100006](https://ec2-staging.gigadb.link/dataset/100006)
Loading failed for the module with source “https://cdn.jsdelivr.net/npm/@loaders.gl/las@4.3.2/+esm”. [100006](https://ec2-staging.gigadb.link/dataset/100006)
Loading failed for the module with source “https://cdn.jsdelivr.net/npm/@loaders.gl/core@4.3.2/+esm”. [100006](https://ec2-staging.gigadb.link/dataset/100006)
Loading failed for the module with source “https://cdn.jsdelivr.net/npm/three@0.170.0/examples/jsm/loaders/STLLoader.js”. [100006](https://ec2-staging.gigadb.link/dataset/100006)
Loading failed for the module with source “https://cdn.jsdelivr.net/npm/three@0.170.0/examples/jsm/loaders/OBJLoader.js”. [100006](https://ec2-staging.gigadb.link/dataset/100006)
Loading failed for the module with source “https://cdn.jsdelivr.net/npm/three@0.170.0/examples/jsm/loaders/PLYLoader.js”. [100006](https://ec2-staging.gigadb.link/dataset/100006)
Loading failed for the module with source “https://cdn.jsdelivr.net/npm/@loaders.gl/las@4.3.2/+esm”. [100006](https://ec2-staging.gigadb.link/dataset/100006)
Loading failed for the module with source “https://cdn.jsdelivr.net/npm/@loaders.gl/core@4.3.2/+esm”.
```
**suggestion:** The fix could be made by adding `https://cdn.jsdelivr.net` in the `script-src` directive in `ops/configuration/nginx-conf/sites/nginx.target_deployment.https.conf.dist`.

kencho51 commented:
Hi @luistoptal,

**issue:**  The `gigadb-website/features` is a legacy dir that waiting for remove.
**suggestion:** The acceptance test should be implemented in `gigadb-website/tests/acceptance/` which would use the codeception framework.
**suggestion:** For the test scenario, usually we would implement in a way that only contains 1 `Given`, 1 `When` and 1 `Then`, which would make the test simpler and easier to test, for example:
```
Scenario: 3D Models tab 
  Given I am not logged in to Gigadb web site
  And I have added "3D Viewer" link "https://s3.ap-northeast-1.wasabisys.com/test-gigadb-datasets/3d-models/100006/GeoB8502_865cm_Shell-4.obj" to dataset "101001"
  When I go to "/dataset/101001"
  Then I should see "3D Models" tab with text "3D Models:"

Scenario: 3D model drop down list
  Given I am not logged in to Gigadb web site
  When I click on the "3D Models" button
  Then I should see a "select.test-model-selector" element
  And I should see a "select.test-model-selector > option" element
  And I should see "GeoB8502_865cm_Shell-4.obj" in the "select.test-model-selector > option" element
```

luistoptal committed:
fix: undoes bug introduced by trim func

luistoptal committed:
chore: add https://cdn.jsdelivr.net to nginx config

luistoptal commented:
@kencho51 I think it would be simpler if one dataset is created already with a 3D model for acceptance testing (one of the S3 links you posted above), what would be the right place to add this? (I see many sql files that are probably used for testing, but not sure which ones are ok and which ones should be deprecated)

kencho51 commented:
Hi @luistoptal,

**suggestion:**

For the test, you could:
1. Add the following line to the file `gigadb-website/data/dev/external_link.csv`:

```
1527,8,https://s3.ap-northeast-1.wasabisys.com/test-gigadb-datasets/3d-models/100006/GeoB8502_865cm_Shell-4.obj,5
```
2. Then spin up gigadb website `./up.sh`, it will execute the migration scripts and update the test database with above entry
3. Then confirm by going to ` http://gigadb.gigasciencejournal.com/dataset/100006`

Then for the implementation, you could add the scenarios to `gigadb-website/tests/acceptance/DatasetView.feature` as

```
  @ok @issue-2054
  Scenario: 3D Models tab
    Given I have not signed in
    When I am on "/dataset/100006"
    Then I should see "3D Models"

  @wip @issue-2054
  Scenario: 3D model drop down list
    Given I have not signed in
    When I follow "3D Models"
    Then I should see "3D Models:"
    And I should see "Select a model"
    And I should see "GeoB8502_865cm_Shell-4.obj"
```

And then you execute the test as below: 

```
% docker-compose run --rm codecept run --no-redirect -g issue-2054 acceptance -v
Codeception PHP Testing Framework v5.0.4 https://helpukrainewin.org
[Groups] issue-2054 

Acceptance Tests (2) --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
Modules: \Helper\Acceptance, Asserts, Db, WebDriver
-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
✔ a user visit the dataset page: 3D Models tab (4.72s)
✔ a user visit the dataset page: 3D model drop down list (3.53s)
-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
Time: 00:25.298, Memory: 12.00 MB

OK (2 tests, 4 assertions)
```

**issue:** Got this error in production `three.module.js:44842 Refused to connect to 'https://s3.ap-northeast-1.wasabisys.com/test-gigadb-datasets/3d-models/100006/GeoB8502_865cm_Shell-4.obj' because it violates the following Content Security Policy directive: "default-src 'self' 'unsafe-inline' 'unsafe-eval' http://gigadb.org http://penguin.genomics.cn https://www.rosaceae.org *.protocols.io https://sketchfab.com https://codeocean.com *.hypothes.is *.datatables.net *.cloudflare.com". Note that 'connect-src' was not explicitly set, so 'default-src' is used as a fallback.`

**suggestion:** The fix could be made by adding `https://s3.ap-northeast-1.wasabisys.com` in the `default-src` directive in `ops/configuration/nginx-conf/sites/nginx.target_deployment.https.conf.dist`.

kencho51 commented:
Hi @luistoptal,

Please use the model files from these links:
```
https://s3.ap-northeast-1.wasabisys.com/gigadb-datasets/dev/tests/3d-models/100006/leaf_01.las
https://s3.ap-northeast-1.wasabisys.com/gigadb-datasets/dev/tests/3d-models/100006/beet_03.stl
https://s3.ap-northeast-1.wasabisys.com/gigadb-datasets/dev/tests/3d-models/100006/GeoB8502_865cm_Shell-4.obj
```

luistoptal committed:
feat: add 3d models to test db and tests

luistoptal committed:
chore: adds model source to nginx

luistoptal commented:
@kencho51 I added your suggestions, that was very helpful

for some reason I had to shut down docker compose and re run up.sh to register the db entires

I also added the wasaby source to all `default-src` entires (all the 3 of them)

luistoptal committed:
chore: Merge branch 'develop' into new-feature/2054-3d-image-viewer

kencho51 commented on features/dataset-view.feature:
```diff
@@ -153,11 +153,15 @@ Feature: a user visit the dataset page
 
 	@ok
-	Scenario:  3D Viewer
+	Scenario: 3D Viewer
 		Given I am not logged in to Gigadb web site
-		And I have added "3D Viewer" link "https://sketchfab.com/models/ea49d0dd500647cbb4b61ad5ca9e659a" to dataset "101001"
+		And I have added "3D Viewer" link "https://s3.ap-northeast-1.wasabisys.com/test-gigadb-datasets/3d-models/100006/GeoB8502_865cm_Shell-4.obj" to dataset "101001"
 		When I go to "/dataset/101001"
 		Then I should see "3D Models" tab with text "3D Models:"
+    When I click on the "3D Models" button
+    Then I should see a "select.test-model-selector" element
+    And I should see a "select.test-model-selector > option" element
+    And I should see "GeoB8502_865cm_Shell-4.obj" in the "select.test-model-selector > option" element
 
```
**suggestion:** This scenario can be removed, since it has been moved to `tests/acceptance/DatasetView.feature` already.

kencho51 commented on sql/3D_Viewer_101001_test_data.sql:
```diff
@@ -1,3 +1,8 @@
 -- to be loaded in production_like.pgdmp
-insert into external_link_type(id, name) values(5, '3D Models');
-insert into external_link(dataset_id, url, external_link_type_id) values(80,'https://sketchfab.com/models/ea49d0dd500647cbb4b61ad5ca9e659a',5);
+insert into external_link_type(id, name)
+select 5, '3D Models'
+where not exists (
+    select 1 from external_link_type
+    where id = 5 and name = '3D Models'
+);
+insert into external_link(dataset_id, url, external_link_type_id) values(80,'https://s3.ap-northeast-1.wasabisys.com/test-gigadb-datasets/3d-models/100006/GeoB8502_865cm_Shell-4.obj',5);
```
**suggestion:** The changes at here should be reverted, as only data from `data/dev` will be used.

luistoptal replied:
I am deleting this file as I think it serves no purpose anymore

kencho51 commented on tests/acceptance/DatasetView.feature:
```diff
@@ -216,4 +216,19 @@ Feature: a user visit the dataset page
+    When I am on "/dataset/100006"
+    Then I should see "3D Models"
+
+  @wip @issue-2054
```
**suggestion:** Change the `@wip` to `@ok` if the scenario is passing, the wip means `work in progress`.

luistoptal committed:
chore: cleanup tests

luistoptal commented on features/bootstrap/DatasetViewContext.php:
```diff
@@ -413,10 +413,7 @@ public function iShouldNotSeeTabWithTable($arg1, TableNode $table)
      */
     public function iHaveAddedLinkToDataset($arg1, $arg2, $arg3)
     {
-        if ("3D Viewer" == $arg1 ) {
```
@kencho51 since I deleted the 3D Models test, I am also deleting this switch case

kencho51 reviewed (approved):
Hi @luistoptal,

With all the requested changes fixed and the 3D models can be displayed in my production, so LGTM, happy to approve.

luistoptal commented:
I noticed a bug I need to fix

luistoptal committed:
fix: displays 3d models only in 3d models tab panel

luistoptal commented:
fixed

luistoptal committed:
fix: render model viewer only once

luistoptal commented:
fixed

rija reviewed (commented):
Hi @luistoptal,

**praise**: @kencho51's remarks were good and I'm glad they were implemented
**praise**: functionality works fine locally and on AWS when using Firefox or a Chromium based browser

**issue**: none of the model loads under Safari (18.3)
there might be a CSP issue for the `.las` model (see screenshot), but the other models won't load either and are not throwing any (meaningful) error in the js console
**issue**: the contextual help "help (h)" and "full screen (f)" appear inline instead of an overlay in Safari

**issue**: the `protected/` directory is a legacy directory for Yii1.1 PHP code, which we will eventually delete once we've [strangled](https://martinfowler.com/bliki/StranglerFigApplication.html) it with other Yii2 and JS apps, 
so we shouldn't put any new app in it (let alone a JS app)
**suggestion**: the javascript code for the functionality should be under `gigadb/app/client`
**chore**: the modal with drop-down menu and contextual hints are used for the first time here it seems, they probably should be added as a specimen in the style guide
**chore**: there are some file conflicts, consider rebasing and fix the conflicts
**praise**: all tests are passing 


<img width="1510" alt="Screenshot 2025-02-07 at 8 00 38 PM" src="https://github.com/user-attachments/assets/caf2166f-d3bd-43a5-a61c-3b818420d249" />

luistoptal committed:
chore: Merge branch 'develop' into new-feature/2054-3d-image-viewer

luistoptal committed:
chore: move model viewer to app folder

luistoptal commented:
@rija thanks for pointing this out. Is there any way I can easily see how the content of the protected folder will be ported to Yii 2 / the new folder structure? I am lost as to what folder structure we want to adopt

I moved the js widget to gigadb/app/client/js. Also, I checked the local deployment in Safari via a cloud crossbrowser testing tool and could see the models fine in Safari 18, the issue you report is from local dev or from staging?

Solved conflicts 👍

rija commented:
reminder to self: 
Manual perform #2200 when this PR is approved

rija commented:
Hi @luistoptal,

There are two type of Yii2 projects, each type having its own directory structure.
The Yii2 advanced template project has the following structure:
https://github.com/yiisoft/yii2-app-advanced

We have examples of such projects under `fuw/app` (that's the File Upload Wizard app) and under `gigareview` (That's the Giga Review app).

There's another type of template, the Yii2 basic template:
https://github.com/yiisoft/yii2-app-basic

that structure is the closest to that of a Yii1.1 project. 

We use the advance template when the application is complex (e.g: it is made of multiple heterogeneous services/integrations, has a back-office components in addition to front-office, has lots of automation and CLI tooling involved)

In any other case, the yii2 basic template should be the default.

Other significant directories in our repo are:

* `gigadb/app/client`: that's where we put all GigaDB project's JS apps (notably FUW's JS is not in there as FUW is considered a different project, same for Gigareview, so they've got their own top-level directory)
* `gigadb/app/models`: the directory for PHP model classes that map to database tables, and could be (must be) shared by any internal PHP applications that need to interact with the Gigadb databases

I don't know if the current main Yii1.1 app will ever be fully transformed into Yii2 app some day, as we tend to work around it until now. But if we ever do it, it will probably be a yii2 basic template project and would probably live under `gigadb/app/server`




Regarding the Safari problems, It happened both locally and on AWS.

I'll try again a bit later this week after I've rebuilt my environment.

@kencho51: have you tried with Safari, did it work for you?

luistoptal committed:
chore: add script to setup gitlab env vars quickly

luistoptal commented:
Hi @rija once I can deploy my fork to staging I will also review different browsers.

> there might be a CSP issue for the .las model (see screenshot)

for some reason I can no longer open the screenshot to a size that allows me to read the logs

I will add the new features to the style guide

luistoptal commented:
@rija Also, I have pushed a commit with a script to update env variables in gitlab. It might not be the best PR for this but for me it all stems from wanting to deploy this branch to production-like to test :) the script does not require any kind of changes to env vars anywhere (.env etc) if it is not used

luistoptal committed:
feat: add 3d model viewer to style guide

luistoptal commented:
Added 3D model viewer to style guide

luistoptal committed:
feat: added env vars file example

pli888 commented:
>I moved the js widget to gigadb/app/client/js. Also, I checked the local deployment in Safari via a cloud crossbrowser testing tool and could see the models fine in Safari 18, the issue you report is from local dev or from staging?

The 3D model viewer functionality works on my `dev`, but there is a problem in my `staging`. When I go to a dataset page with a dataset that has a 3D model external link, I see a `The asset "/gigadb/app/client/js/model-viewer" to be published does not exist` error message. See https://gigadb-staging-mumbai.gigatools.net/dataset/100295.

kencho51 commented:
Hi @rija 

> @kencho51: have you tried with Safari, did it work for you?

It works on my Safari with version 18.3 (20620.2.4.11.5).

rija commented:
Hi @luistoptal 

**praise**: I saw that the 3D models tab is now in the style guide
**praise**: I've retested locally this PR, And all three models load correctly on Safari on my local dev.

(My pipeline is still building, I'll let you know the results from my AWS deployment later)

**issue**: However the contextual help and full screen callbacks still appear inline on Safari (including the example one from the style guide)

luistoptal commented:
HI @rija 

> contextual help and full screen callbacks still appear inline on Safari

Can you plase add screenshots so I understand better what you see?

rija commented:
> HI @rija
> 
> > contextual help and full screen callbacks still appear inline on Safari
> 
> Can you plase add screenshots so I understand better what you see?

Hi @luistoptal 
don't worry, ignore my remark as after hard reset of the cache, it behaves correctly locally on Safari on the dataset page and on the style guide.

I'll let you know when I get your PR deployed on AWS

rija reviewed (commented):
Hi @luistoptal,

**issue**: Having deployed your PR to AWS, I'm now seeing the same problem as @pli888:

CException:
```
The asset "/gigadb/app/client/js/model-viewer" to be published does not exist.
```

Neither the web server nor the application server knows about that path, and they probably shouldn't anyway.

**suggestion**: can you pre-compile the model-viewer code into the top-level `js/` directory (it is effectly our `dist` directory), so that is can be copied into both the web and application containers? 
You would need to load the javascript in the page from that directory.
That's the approach we took for the File Upload Wizard. (that's why you can see `fuw-1.0.0.js` in the `js` directory, while the javascript source is under `gigadb/app/client/web`)

luistoptal commented:
Hi @rija

So if I understand correctly, the folder /gigadb/app/client/js/model-viewer is not available in the dockerized production app?

will the following work?

```php
$jsDir = Yii::getAlias('js/model-viewer'); // using the js folder instead of the gigadb folder
$jsUrl = Yii::app()->assetManager->publish($jsDir); // this will create a dir under the assets folder

Yii::app()->clientScript->registerScriptFile($jsUrl . '/index.js', CClientScript::POS_END, ['type' => 'module']);
?>

<script type="module">
  import { modelViewer } from "<?php echo $jsUrl; ?>/index.js";

  // code
</script>
```

or must I call the script directly from the js folder without using the php helpers? in which case, I do not understand how the assets folder works

rija commented:
Hi @luistoptal,

I understand your confusion, sorry about that.
I forgot we were trying to use Yii assets mechanism here.

First, do not use the `js/` directory for the javascript source code (although, to answer your question, it would fix the error) as it is not meant for that, it is used  for the same purpose as the Yii assets directory (because I wasn't familiar with Yii assets when I was doing FUW).

You are right in realising that the problem is that the path `/gigadb/app/client/js/model-viewer` (which is the right place for the javascript source code to be) is not available in the in the dockerized production app.

So, given that you are using Yii assets, the correct fix is to make that path available to the dockerized production app.

We do so by copying the path into the Dockerfile for the production Web and application container by adding the following line:
```
COPY ./gigadb/app/client/js /gigadb/app/client/js
```
into `ops/packaging/Production-Web-Dockerfile` and `ops/packaging/Production-Dockerfile`

in the same section were the other `COPY` instructions are.

luistoptal committed:
fix: make viewer js folder available to prod container

luistoptal commented:
Hi @rija I added the copy command to the docker files as per your suggestion, thanks

pli888 commented:
@luistoptal Seems like there are conflicts with `CHANGELOG.md` and `datatables.less`. Can you rebase with latest develop branch code then I can test this PR again?

rija reviewed (commented):
Hi @luistoptal,

**praise**: Your last change to the dockerfiles fixed the asset publishing error, thanks.

**issue**: when loading the '.las' model (on any browser) I get the following CSP error the js console:

```
Content-Security-Policy: The page's settings blocked a worker script (worker-src) at blob:https://rija-staging.gigadb.host/ff12fa1f-6144-4fa4-a046-e14c8902478c from being executed because it violates the following directive: "child-src <https://www.protocols.io>](https://www.protocols.io) <https://hypothes.is>](https://hypothes.is) <https://www.rosaceae.org>](https://www.rosaceae.org) <https://sketchfab.com>](https://sketchfab.com) <https://codeocean.com>](https://codeocean.com) <https://tumormap.ucsc.edu>](https://tumormap.ucsc.edu)"
```

and the model-viewer shows that error:
```
Error: Failed to load worker las (#1 of 3) from https://unpkg.com/@loaders.gl/las@4.3.1/dist/las-worker.js.
```
suggestion: It sounds like there is a missing directive, maybe you could add a 
```
; worker-src 'self' blob:
```
at the end (but within the double-quote) of the two lines starting with `add_header                Content-Security-Policy` and `add_header                Content-Security-Policy-Report-Only` respectively.

and I think we may also need to add `https://unpkg.com` to the `script-src` directive on both lines as it is not listed there.

luistoptal committed:
chore: Merge branch 'develop' into new-feature/2054-3d-image-viewer

luistoptal committed:
fix: allows 3d model load related scripts and workers in nginx server

luistoptal commented:
Hi @rija I updated the nginx server config as you suggested

rija reviewed (commented):
Hi @luistoptal,

**praise**:  the functionality now works for me in all web browsers
**praise**: I can see how `ops/scripts/set_env_vars.sh` can be useful and it's great a mention of it has been added in the docs
**nitpick**: I feel like the `.gitlab-env-vars.example` example file should go in `ops/configuration/variables` with the other variables related example/sample files

luistoptal committed:
chore: relocate gitlab env var example file

luistoptal commented:
Hi @rija I relocated the example file as you suggested

rija reviewed (approved):
Hi @luistoptal,

based on the latest change, I'm happy to approve this PR

kencho51 commented:
Hi @luis,

>Also, I have pushed a commit with a script to update env variables in gitlab. It might not be the best PR for this but for me it all stems from wanting to deploy this branch to production-like to test :) the script does not require any kind of changes to env vars anywhere (.env etc) if it is not used

Regarding this script which aims to set up variables for staging and live deployment:
**question:(non-blocking)** In order to execute this script, I have to manually update the `GITLAB_PROJECT_ID` and `GITLAB_PRIVATE_TOKEN` in the .secrets and also copy `.gitlab-env-vars.example` to `.gitlab-env-vars`?
**suggestion: (non-blocking)** Maybe create a block in `SETUP_CI_CD_PIPELINE.md`, for example, as below:

```
% cd gigadb-website
% ./up.sh 
% cp ops/configuration/variables/gitlab-env-vars.example .gitlab-env-vars
% ./ops/scripts/set_env_vars.sh
```

**suggestion:(non-blocking)** For var `GITLAB_PRIVATE_TOKEN`, you can get it from `.env`, while for the var `GITLAB_PROJECT_ID`, can it be replaced by `PROJECT_VARIABLES_URL` which is also in .env.

luistoptal committed:
chore: use env vars from env file

luistoptal commented:
@kencho51 I applied your suggestions, thanks

pli888 reviewed (approved):
**praise**: Functionality is now working on my staging and dev environments.
**praise**: New code looks good including new `set_env_vars.sh` script.

Happy to approve.

rija committed:
Merge branch 'develop' into new-feature/2054-3d-image-viewer

rija commented:
when deploying, will need to run the following query:
```
insert into external_link (dataset_id, url, external_link_type_id) select dataset_id, location as url, 5 as external_link_type_id from file where location like '%.stl' or location like '%.obj' or location like '%.las' or location like '%.ply';
```
